File: 1520-NumberOfStepsToReduceANumberInBinaryRepresentationToOne/1520-NumberOfStepsToReduceANumberInBinaryRepresentationToOne.cpp
// Last updated: 06/07/2026, 19:03:23
class Solution {
public:
    int numSteps(string s) {
        int count = 0;

        while (s != "1") {
            if (s.back() == '0') {
                s.pop_back();
            } else {
                int i = s.size() - 1;

                while (i >= 0 && s[i] == '1') {
                    s[i] = '0';
                    i--;
                }

                if (i < 0) {
                    s = "1" + s;
                } else {
                    s[i] = '1';
                }
            }

            count++;
        }

        return count;
    }
};

File: 1520-NumberOfStepsToReduceANumberInBinaryRepresentationToOne/1520-NumberOfStepsToReduceANumberInBinaryRepresentationToOne.py
# Last updated: 06/07/2026, 19:03:28
class Solution:
    def numSteps(self, s: str) -> int:
        if (s == "1"):
            return 0

        if (s[-1] == "0"):
            return 1 + self.numSteps(s[:-1])
        else:
            chars = [ch for ch in s]
            chars[-1] = "0"
            j = len(s) - 2
            while (j >= 0):
                if (chars[j] == "0"):
                    chars[j] = "1"
                    break
                else:
                    chars[j] = "0"
                j -= 1

            if (j == -1):
                chars = list("1") + chars
                

            return 1 + self.numSteps(''.join(chars))